src/libsyntax/test.rs

```diff
@@ -123,41 +119,21 @@ impl<'a> fold::Folder for TestHarnessGenerator<'a> {
         }
         debug!("current path: {}", path_name_i(&self.cx.path));
 
-        if is_test_fn(&self.cx, &i) || is_bench_fn(&self.cx, &i) {
-            match i.node {
-                ast::ItemKind::Fn(_, header, _, _) => {
-                    if header.unsafety == ast::Unsafety::Unsafe {
-                        let diag = self.cx.span_diagnostic;
-                        diag.span_fatal(
-                            i.span,
-                            "unsafe functions cannot be used for tests"
-                        ).raise();
-                    }
-                    if header.asyncness.is_async() {
-                        let diag = self.cx.span_diagnostic;
-                        diag.span_fatal(
-                            i.span,
-                            "async functions cannot be used for tests"
-                        ).raise();
-                    }
-                }
-                _ => {},
-            }
-
+        let mut item = i.into_inner();
+        if is_test_case(&item) {
             debug!("this is a test function");
+
             let test = Test {
-                span: i.span,
+                span: item.span,
                 path: self.cx.path.clone(),
-                bench: is_bench_fn(&self.cx, &i),
-                ignore: is_ignored(&i),
-                should_panic: should_panic(&i, &self.cx),
-                allow_fail: is_allowed_fail(&i),
             };
-            self.cx.testfns.push(test);
-            self.tests.push(i.ident);
+            self.cx.test_cases.push(test);
+            self.tests.push(item.ident);
+
+            // Ensure this item is capable of being reexported
+            item.vis = respan(item.vis.span, ast::VisibilityKind::Public);
```

Comment:
This seems slightly suspect to me in the sense that it messes with the module system in a user-visible fashion, but I haven't finished reading this PR yet so I'll keep going!

Reply:
Ah this is actually a mistake. This is already handled in `expand.rs` where we both gensym and make public so it's not user-visible. This line essentially a no-op currently.